pkg/kubelet/util/pluginwatcher/plugin_watcher_test.go

```diff
@@ -199,6 +103,79 @@ func TestExamplePlugin(t *testing.T) {
 	select {
 	case err = <-errCh:
 		require.NoError(t, err)
+	case <-chanForHandlerAckTesting:
+		t.Fatalf("handler wait for notification ack error")
+	case <-time.After(time.Second):
+		t.Fatalf("Timed out while waiting for watcher start")
+
+	}
+
+	require.NoError(t, w.Stop())
+	err = w.Cleanup()
+	require.NoError(t, err)
+}
+
+func TestPluginWithSubDir(t *testing.T) {
+	rootDir, err := ioutil.TempDir("", "plugin_test")
+	require.NoError(t, err)
+
+	w := NewWatcher(rootDir)
+
+	w.AddHandler(registerapi.CSIPlugin, exampleHandler)
+	w.AddHandler(registerapi.DevicePlugin, exampleHandler)
+
+	err = w.fs.MkdirAll(filepath.Join(rootDir, registerapi.DevicePlugin), 0755)
+	require.NoError(t, err)
+	err = w.fs.MkdirAll(filepath.Join(rootDir, registerapi.CSIPlugin), 0755)
+	require.NoError(t, err)
+
+	dpSocketPath := filepath.Join(rootDir, registerapi.DevicePlugin, "plugin.sock")
+	csiSocketPath := filepath.Join(rootDir, registerapi.CSIPlugin, "plugin.sock")
+
+	require.NoError(t, w.Start())
+
+	// two plugins using the same name but with different type
+	dp := NewTestExamplePlugin("exampleplugin", registerapi.DevicePlugin, "example-endpoint")
+	require.NoError(t, dp.Serve(dpSocketPath))
+	require.True(t, waitForPluginRegistrationStatus(t, dp.registrationStatus))
+
+	csi := NewTestExamplePlugin("exampleplugin", registerapi.CSIPlugin, "example-endpoint")
+	require.NoError(t, csi.Serve(csiSocketPath))
+	require.True(t, waitForPluginRegistrationStatus(t, csi.registrationStatus))
+
+	// Restarts plugin watcher should traverse the socket directory and issues a
+	// callback for every existing socket.
+	require.NoError(t, w.Stop())
+	errCh := make(chan error)
+	go func() {
+		errCh <- w.Start()
+	}()
+
+	var wg sync.WaitGroup
+	wg.Add(2)
+	var dpStatus string
+	var csiStatus string
+	go func() {
+		dpStatus = strconv.FormatBool(waitForPluginRegistrationStatus(t, dp.registrationStatus))
```

Comment:
Why does this need to be in a goroutine?

Reply:
it is just a way to wait two events without order, the alternative would be: wait one first, and wait another which is not optimal.

Reply:
sounds good